### PR TITLE
HBHE M2: noise definition cleanup

### PR DESCRIPTION
**Pull request by mariadalfonso**
Redefined some of the noise terms in the M2chi2.
Mainly to take values from the DB and not from the hard coded python.
Only minor changes are expected in the Phase1-siPM
Validation slides are here
http://dalfonso.web.cern.ch/dalfonso/NoiseDefCleanup.pdf

### CODE REVIEW

**Comment by slava77** [comment]
@mariadalfonso 
I thought we agreed to keep HPD reco in sync (either change it simultaneously in phase-0 reco or not change at all).
The short term goal for the HPD reco would be to switch to using the phase-1 reconstructor for all data.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by mariadalfonso** [comment]
@slava77 
Rolled back the python->DB changes for HPD. 
kept comments in the code so that we can come back to this at the very end.
PR Description update

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@mariadalfonso 

(I should have checked at jenkins level much earlier)
there are differences in run1 and run2 workflows
e.g. 25.0 (ttbar no PU from run1)
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_9_0_X_2017-01-09-1100+17129/17793/25.0_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT/HcalRecHitsD_HcalRecHitTask.html

Most notably, chi2 at 999 have moved to chi2=0 (or is it -9999 which just shows up as 0 in root's log10).

File: RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc [hunk 1]
```diff
@@ -344,15 +346,7 @@ void PulseShapeFitOOTPileupCorrection::apply(const CaloSamples & cs,
...
    if(tstrig >= ts4Min_&& tsTOTen > 0.) { //Two sigma from 0
-     pulseShapeFit(energyArr, pedenArr, chargeArr, pedArr, gainArr, tsTOTen, fitParsVec, noiseADCArr);
-   }
-   else if((tstrig < ts4Min_||tsTOTen < 0.)&&(ts4Min_==0)){
-     fitParsVec.clear();
-     fitParsVec.push_back(0.);
-     fitParsVec.push_back(0.);
-     fitParsVec.push_back(0.);
-     fitParsVec.push_back(999.);
-     fitParsVec.push_back(false);
```
Comment (slava77) [inline RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc]: it looks like removal of this ```else if ``` is actually triggering some small performance changes.
E.g. in 136.731 workflow about 10% of the hits were in this branch; so, their time and chi2 values have changed

File: RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc [hunk 2]
```diff
@@ -382,7 +376,7 @@ int PulseShapeFitOOTPileupCorrection::pulseShapeFit(const double * energyArr, co
       tmpslew[i] = 0;
       if(applyTimeSlew_) tmpslew[i] = HcalTimeSlew::delay(std::max(1.0,chargeArr[i]),slewFlavor_); 
       // add the noise components
-      tmperry2[i]=noise_*noise_+ noiseADCArr[i]*noiseADCArr[i];
+      tmperry2[i]=noiseArr[i]*noiseArr[i];
```
Comment (slava77) [inline RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc]: it looks like noiseArr argument in pulseShapeFit is used only squared.
Also, it looks like the noiseArr derivation in the apply method is first computed as a square of the noise and sqrt is taken for no particular reason.
I suggest to pass noiseArr2 and eliminate the "sqr(sqrt" cycle and not waste CPU.
I guess this is also the origin of the small discrepancies in energy in pre-phase-1 workflows.

**Comment by slava77** [comment]
an example of changes in 136.731 (from jenkins comparisons)

the significant change in chi2 default value:
all_oldvsnew_runsingleph2016bwf136p731c_log10hbherechitssorted_hbheprereco_rereco_obj_obj_chi2
If I understand correctly, it still comes out with the same energy=0.

There are also some small variations in the chi2 bulk with small changes in energy


HB energy (plot for E>1e-3) 
all_oldvsnew_runsingleph2016bwf136p731c_log10hbherechitssorted_hbhereco_rereco_obj_obj_energy3
HE energy 
all_oldvsnew_runsingleph2016bwf136p731c_log10hbherechitssorted_hbhereco_rereco_obj_obj_energy2

this looks pretty small

File: RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc (hunk 1 above)
Comment (mariadalfonso) [inline RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc]: @slava77 
the chi2 was only introduced in the 81X wasn't there in the run2-2016 #15916
so it's ok that we have just pick one value -9999 for all the failing

File: RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc (hunk 2 above)
Comment (mariadalfonso) [inline RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc]: @slava77 ok I can remove the sqrt and square step

File: RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc (hunk 1 above)
Comment (slava77) [inline RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc]: ok, makes sense for chi2.

default time value of -9999 is probably OK (except for maybe some code that tried to make sense of the two options for E=0)

File: RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc [hunk 3]
```diff
@@ -330,6 +331,7 @@ void PulseShapeFitOOTPileupCorrection::apply(const CaloSamples & cs,
       energyArr[ip] = energy; pedenArr[ip] = peden;
 
       noiseADCArr[ip] = psfPtr_->sigmaHPDQIE8(chargeArr[ip]); // Add Greg's channel discretization
+      noiseArr[ip] = noise_ * noise_ + noiseADCArr[ip] * noiseADCArr[ip];
```
Comment (slava77) [inline RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc]: please change to noiseArrSq to reduce possible confusion

**Comment by mariadalfonso** [comment]
@slava77 
done fixes from my side  (thanks for spotting them).

The HB energy should not change at all.
The HE energy is expected to be very small, as in practice only the pedestal gaussian constraint change. This PR mainly remove the dependency from the hard coded python, so that any future change of the noise/gains/responses from data online/offline will be transparent.

**Comment by slava77** [comment]
@cmsbuild please test

@mariadalfonso thank you for the quick turnaround

**Comment by slava77** [comment]
@cmsbuild please test

@smuzaffar 
in the last comparisons run it looks like all validateJR failed with 
``` cc1plus: error:  ... permission denied ```
can this be detected somehow?

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/17129		fbd50e6
- code changes are in line with the description and the follow up review
- jenkins tests pass and comparisons with baseline show differences in expected places: 
    - phase-0 (HCAL perspective) has changes only in the default values of chi2 and time
    - phase-1 (2017 and 2023 wfs) show minor changes in HE which propagate to minor changes downstream
- local tests using CMSSW_9_0_X_2017-01-11-2300 as a baseline
show that changes are roughly as expected from slightly different (and effectively smaller due to addition in quadrature) noise component in the minimization in HE for 2017 and in HB for 2023 (SiPM channels):
    - CPU time is up by ~5% (0.8% of the total reco time)
    - chi2 is up slightly 
    - low energy hit reconstruction has some noticeable variations (without a significant pattern)

Also, there are previously noted changes in HB hit time and chi2 values corresponding to defaults.

Pi E 5 to 5000 (low energy is mostly noise)
all_sign822vsorig_pie5to5000in2017c_log10hbherechitssorted_hbhereco__reco_obj_obj_energy13

ttbar PU35 (10224)
log10(energy) in HE for E>0.001
all_sign822vsorig_ttbar13tevpu2017wf10224p0c_log10hbherechitssorted_hbhereco__reco_obj_obj_energy13


log10(chi2) in HE
all_sign822vsorig_ttbar13tevpu2017wf10224p0c_log10hbherechitssorted_hbhereco__reco_obj_obj_chi220

The effect on higher level objects is fairly small: jet response is essentially the same (within ~0.5%)

**Comment by davidlange6** [comment]
+1

**Pull request merged**